File: kata/bowling_game/2021_03_30/src/game.py
class Game:
    def __init__(self):
        self.__rolls = [0] * 21
        self.__current_roll = 0

    def roll(self, pins):
        self.__rolls[self.__current_roll] = pins
        self.__current_roll += 1

    def score(self):
        score = 0
        frame_index = 0
        for frame in range(10):
            if self.__is_strike(frame_index):
                score += 10 + self.__strike_bonus(frame_index)
                frame_index += 1
            elif self.__is_spare(frame_index):
                score += 10 + self.__spare_bonus(frame_index)
                frame_index += 2
            else:
                score += self.__sum_of_balls_in_frame(frame_index)
                frame_index += 2

        return score

    def __is_strike(self, frame_index):
        return self.__rolls[frame_index] == 10

    def __sum_of_balls_in_frame(self, frame_index):
        return self.__rolls[frame_index] + self.__rolls[frame_index + 1]

    def __spare_bonus(self, frame_index):
        return self.__rolls[frame_index + 2]

    def __strike_bonus(self, frame_index):
        return self.__rolls[frame_index + 1] + self.__rolls[frame_index + 2]

    def __is_spare(self, frame_index):
        return self.__rolls[frame_index] + self.__rolls[frame_index + 1] == 10

File: kata/bowling_game/cpp/2024_01_09/src/Game.cpp
#include "Game.h"

void Game::roll(int pins) {
  score += pins;
  rolls[currentRoll++] = pins;
}

int Game::getScore() {
  int score = 0;
  int frameIndex = 0;
  for (int frame = 0; frame < 10; frame++) {
    if (isStrike(frameIndex)) {
      score += strikeBonus(frameIndex);
      frameIndex += 1;
    } else if (isSpare(frameIndex)) {
      score += spareBonus(frameIndex);
      frameIndex += 2;
    } else {
      score += sumOfBallsInFrame(frameIndex);
      frameIndex += 2;
    }
  }

  return score;
}

bool Game::isSpare(int frameIndex) {
  return rolls[frameIndex] + rolls[frameIndex + 1] == 10;
}

bool Game::isStrike(int frameIndex) { return rolls[frameIndex] == 10; }

int Game::sumOfBallsInFrame(int frameIndex) {
  return rolls[frameIndex] + rolls[frameIndex + 1];
}

int Game::spareBonus(int frameIndex) { return 10 + rolls[frameIndex + 2]; }

int Game::strikeBonus(int frameIndex) {
  return 10 + rolls[frameIndex + 1] + rolls[frameIndex + 2];
}
